### PR TITLE
PIA-625: Introduce DNS option to enable system resolver

### PR DESCRIPTION
**Pull request by kp-juan-docal**
## Summary

Introduce a new DNS option called `System DNS Resolver` with the intention of allowing the use of Android's Privacy DNS setting.

@kp-iva-nedeleva we are waiting for translations, but it is ready to be reviewed in the meantime.

## Sanity Tests
With Android Private DNS set to `one.one.one.one`

### Wireguard
- [x] Go to Settings, Network, DNS, and Enable `System DNS Resolver`. Confirm it prompts for Allowing LAN access.
- [x] After the above. Connect to any region. Confirm it connects successfully.
- [x] After the above. Open `1.1.1.1/help`. Confirm we are set as using the DNS and with some level of privacy (e.g. DoT, DoH).
- [x] After the above. Open `dnsleak.com`. Confirm it shows Cloudflare as ISP.

### OpenVPN
- [x] Go to Settings, Network, DNS, and Enable `System DNS Resolver`. Confirm it prompts for Allowing LAN access.
- [x] After the above. Connect to any region. Confirm it connects successfully.
- [x] After the above. Open `1.1.1.1/help`. Confirm we are set as using the DNS and with some level of privacy (e.g. DoT, DoH).
- [x] After the above. Open `dnsleak.com`. Confirm it shows Cloudflare as ISP.

### CODE REVIEW

**Review by kp-iva-nedeleva — DISMISSED** [review]
LGTM!

**Review by kp-iva-nedeleva — DISMISSED** [review]
LGTM!

File: app/src/main/java/com/privateinternetaccess/android/wireguard/backend/GoBackend.java [hunk 1]
```diff
@@ -310,8 +349,9 @@ private void setStateInternal(
             }
 
             for (final Peer peer : config.getPeers()) {
-                for (final InetNetwork addr : peer.getAllowedIps())
+                for (final InetNetwork addr : peer.getAllowedIps()) {
```
Comment (kp-aldo-pedromingo) [inline app/src/main/java/com/privateinternetaccess/android/wireguard/backend/GoBackend.java:352]: Thank you!

File: app/src/main/java/com/privateinternetaccess/android/utils/SubnetUtils.kt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+        operator fun plus(value: Int): IpAddress = IpAddress(intValue + value)
+        override fun compareTo(other: IpAddress): Int = intValue.compareTo(other.intValue)
+        override fun toString(): String = "${(intValue ushr 24) and 0xFF}.${(intValue ushr 16) and 0xFF}.${(intValue ushr 8) and 0xFF}.${intValue and 0xFF}"
+    }
+
+    private data class IpSubnet(val base: IpAddress, val mask: Int) {
+        fun contains(address: IpAddress): Boolean {
+            val maskBits = (1 shl mask) - 1 shl (32 - mask)
+            return (base.intValue and maskBits) == (address.intValue and maskBits)
```
Comment (kp-aldo-pedromingo) [inline app/src/main/java/com/privateinternetaccess/android/utils/SubnetUtils.kt:51]: Please add parenthesis to make the operation more explicit:
((1 shl mask) - 1) shl (32 - mask)

**Review by kp-iva-nedeleva — APPROVED** [review]
LGTM!

**Pull request merged**